common/logSanitizer.go

```diff
@@ -0,0 +1,103 @@
+// Copyright © Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"github.com/Azure/azure-pipeline-go/pipeline"
+	"regexp"
+	"strings"
+)
+
+// azCopyLogSanitizer performs string-replacement based log redaction
+// This serves as a backstop, to help make sure that secrets don't get logged.
+// It does search and replace of the types of credentials that are expected to exist in an
+// azCopy application (which are typically auth tokens inside URLs).
+// While we already have code that redacts these from known request headers and request URLs,
+// they can still be put into errors (e.g. HTTP errors) and if those errors are logged then those
+// secrets will leak into the logs if we don't have this class to filter them out.
+// The alternative would be to filter at all sites where such errors (and other things) may be
+// logged, but that's less maintainable in the long term.
+type azCopyLogSanitizer struct {
+}
+
+func NewAzCopyLogSanitizer() pipeline.LogSanitizer {
+	return &azCopyLogSanitizer{}
+}
+
+var sensitiveQueryStringKeys = []string{
+	strings.ToLower(SigAzure),
```

Comment:
should a "=" follow the key? If we look for all "sig"s, we might be redacting things other than secrets, for example, if the user's account is called `signaturevent` or something like that.

Reply:
If I add a test, as per your other comment (and ensure the test passes) does that cover this comment too?

Reply:
Actually, it will already pass. (And it does, I just added it). Reason is that the `Contains` test will think its a match, but the regex test will not, so nothing will be replaced.  So only downside of not having the = in the "key string" is the extra perf hit of running the regex -but that only happens if the `Contains` gets a match - and no most strings it won't so we won't have the perf hit.

The upside of _not_ having the = here is that it gives us flexibility to successfully redact things that are not as strictly formatted as query strings. Specifically, things where there is as space before the = or there's a : instead of a =.  I'm not sure that we realistically expect to encounter any of those in the near future; but doing it the way its don't now at least does offer that flexibility (and I can't really see a downside, since the perf hit, as above, will be small and very rare).

Reply:
I propose to mark this conversation resolved, but if I'm missing something please reopen it.

Reply:
Ok, the reasoning makes sense to me.